buildSrc/subprojects/parser/src/test/java/org/apache/calcite/test/DialectGenerateTest.java

```diff
@@ -0,0 +1,158 @@
+    String foo = "void foo() :\n"
+      + "{\n"
+      + "    String x = \" '}' \";\n"
+      + "}\n"
+      + "{\n"
+      + "    // overidden by testDialect\n"
+      + "}";
+    String bar = "void bar  (   )  : {} {}";
+    String baz = "void baz() : {x}\n"
+      + "{\n"
+      + "    // overridden by intermediate\n"
+      + "}";
+    String qux = "void qux(int arg1, int arg2) :\n"
+      + "{\n"
+      + "    String x = \" } \"\n"
+      + "}\n"
+      + "{\n"
+      + "    // Below is a }\n"
+      + "}";
+    String quux = "void quux(int arg1,\n"
+      + "    int arg2\n"
+      + ") :\n"
+      + "{\n"
+      + "    char x = ' } '\n"
+      + "    String y;\n"
+      + "}\n"
+      + "{\n"
+      + "    /* All invalid curly braces:\n"
+      + "    }\n"
+      + "    // }\n"
+      + "    \" } \"\n"
+      + "    ' } '\n"
+      + "    */\n"
+      + "    <TOKEN> {\n"
+      + "        y = \" { } } \"\n"
+      + "    }\n\n"
+      + "    // Not a string: \"\n"
+      + "}";
+    return new LinkedHashMap<String, String>() {{
+      put("foo", foo);
+      put("bar", bar);
+      put("baz", baz);
+      put("qux", qux);
+      put("quux", quux);
+    }};
+  }
+
+  private Queue<String> getTokens(String str) {
+    return new LinkedList(Arrays.asList(DialectGenerate.tokenizerPattern.split(str)));
+  }
+
+  private void assertCurlyParserSucceeds(String input) {
+    CurlyParser parser = new CurlyParser();
+    Queue<String> tokens = getTokens(input);
+    boolean doneParsing = false;
+    while (!tokens.isEmpty()) {
+      if (doneParsing) {
+        assertTrue(false, "Failed to parse curly block");
+      }
+      doneParsing = parser.parseToken(tokens.poll());
+    }
+    assertTrue(doneParsing);
+  }
+
+  @Test public void testCurlyParserParsesEmpty() {
+    String input = "{  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testCurlyParserParsesString() {
+    String input = "{ \" } \"  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testCurlyParserParsesCharacter() {
+    String input = "{ ' } '  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testCurlyParserParsesSingleComment() {
+    String input = "{ //  } \n  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testCurlyParserParsesMultiComment() {
+    String input = "{ /*  } */  }";
+    assertCurlyParserSucceeds(input);
+  }
+
+  @Test public void testExtractFunctions() {
+    DialectGenerate dialectGenerate = setupDialectGenerate();
+    Map<String, String> res = dialectGenerate.extractFunctions();
+    Map<String, String> expected = getExpectedExtractionResults();
+    assertEquals(expected.size(), res.size(),
+        "Resultant map size doesn't match expected map size");
+    for (String key : expected.keySet()) {
+      assertTrue(res.containsKey(key), "Resultant map doesn't contain expected key: " + key);
+      assertEquals(expected.get(key), res.get(key));
+    }
+  }
+}
```

Comment:
In addition to the CurlyParser unit tests and the testExtractFunctions end-to-end test, did we want to add any unit tests for DialectGenerate's processFile and/or processFunction?